cpp/test/core/maybe_tests.cpp

```diff
@@ -107,6 +126,259 @@ TEST_CASE_BEGIN(MaybeInterface)
+
+    { // type with allocator
+        auto numbers = { 1, 2, 3, 4, 5 };
+        auto primes = { 2, 3, 5 };
+        std::vector<int> vecNumbers{ std::begin(numbers), std::end(numbers) };
+        std::vector<int> vecPrimes{ std::begin(primes), std::end(primes) };
+
+        bond::maybe<std::vector<int>> copiedValue{ vecNumbers };
+        BOOST_REQUIRE(copiedValue);
+        BOOST_CHECK(copiedValue.value() == vecNumbers);
+
+        copiedValue = vecPrimes;
+        BOOST_REQUIRE(copiedValue);
+        BOOST_CHECK(copiedValue.value() == vecPrimes);
+
+        bond::maybe<std::vector<int>> movedValue{ std::move(vecNumbers) };
+        BOOST_CHECK(vecNumbers.empty());
+        BOOST_REQUIRE(movedValue);
+        BOOST_CHECK(movedValue.value().size() == 5);
+
+        movedValue = std::move(vecPrimes);
+        BOOST_CHECK(vecPrimes.empty());
+        BOOST_REQUIRE(movedValue);
+        BOOST_CHECK(movedValue.value().size() == 3);
+    }
+
+    { // move-only type with allocator
+        std::vector<std::unique_ptr<int>> source1;
+        source1.emplace_back(new int{ 1 });
+
+        bond::maybe<std::vector<std::unique_ptr<int>>> movedValue{ std::move(source1) };
+        BOOST_CHECK(source1.empty());
+        BOOST_REQUIRE(movedValue);
+        BOOST_CHECK(movedValue.value().size() == 1);
+
+        std::vector<std::unique_ptr<int>> source2;
+        source2.emplace_back(new int{ 2 });
+        source2.emplace_back(new int{ 2 });
+
+        movedValue = std::move(source2);
+        BOOST_CHECK(source2.empty());
+        BOOST_REQUIRE(movedValue);
+        BOOST_CHECK(movedValue.value().size() == 2);
+    }
+}
+TEST_CASE_END
+
+TEST_CASE_BEGIN(MovedFromIsNothing)
+{
+    { // simple type
+        bond::maybe<bool> m{ false };
+        BOOST_REQUIRE(!m.is_nothing());
+
+        bond::maybe<bool> sink{ std::move(m) };
+        BOOST_REQUIRE(m.is_nothing());
+        BOOST_REQUIRE(!sink.is_nothing());
+        BOOST_CHECK(!sink.value());
+    }
+
+    { // move-only type
+        bond::maybe<std::unique_ptr<int>> moveOnly;
+        moveOnly.emplace(new int{ 42 });
+        BOOST_REQUIRE(!moveOnly.is_nothing());
+
+        bond::maybe<std::unique_ptr<int>> moveOnlySink{ std::move(moveOnly) };
+        BOOST_REQUIRE(moveOnly.is_nothing());
+        BOOST_REQUIRE(!moveOnlySink.is_nothing());
+        BOOST_CHECK(*moveOnlySink.value() == 42);
+    }
+
+    { // move-only type with allocator
+        bond::maybe<std::vector<std::unique_ptr<int>>> m;
+        m.emplace();
+        m.value().emplace_back(new int{ 1 });
+        m.value().emplace_back(new int{ 2 });
+        BOOST_REQUIRE(!m.is_nothing());
+
+        bond::maybe<std::vector<std::unique_ptr<int>>> sink{ std::move(m) };
+        BOOST_REQUIRE(m.is_nothing());
+        BOOST_REQUIRE(!sink.is_nothing());
+        BOOST_CHECK(sink.value().size() == 2);
+    }
+}
+TEST_CASE_END
+
+struct UsesAllocator
+{
+    using allocator_type = allocator_with_state<>;
+
+    UsesAllocator() = default;
```

Comment:
Also a test with non-default-constructible allocator (perhaps skipping VC12).

Reply:
`allocator_with_state` is not default constructible, so I'll just remove this.